search/services/index/highlighting.py

```diff
@@ -0,0 +1,269 @@
+
+    Parameters
+    ----------
+    result : dict
+        Contains processed search result data destined for the caller.
+    raw : :class:`.Response`
+        A response from Elasticsearch.
+
+    Returns
+    -------
+    dict
+        The ``result`` object, updated with ``highlight`` and ``preview``
+        items.
+
+    """
+    if not hasattr(raw.meta, 'highlight'):
+        return result   # Nothing to do.
+
+    result['highlight'] = {}
+    # The values here will (almost) always be list-like. So we need to stitch
+    # them together.
+    for field in dir(raw.meta.highlight):
+        value = getattr(raw.meta.highlight, field)
+        if hasattr(value, '__iter__'):
+            value = '&hellip;'.join(value)
+
+        # Non-TeX searches may hit inside of TeXisms. Highlighting those
+        # fragments (i.e. inserting HTML) will break MathJax rendering.
+        # To guard against this while preserving highlighting, we move
+        # any highlighting tags from within TeXisms to encapsulate the
+        # entire TeXism.
+        if field in ['title', 'title.english',
+                     'abstract', 'abstract.english']:
+            value = _highlight_whole_texism(value)
+
+        # A hit on authors may originate in several different fields, most
+        # of which are not displayed. And in any case, author names may be
+        # truncated. So instead of highlighting author names themselves, we
+        # set a 'flag' that can get picked up in the template and highlight
+        # the entire author field.
+        if field.startswith('author'):
+            field = 'author'
+            value = True
+        result['highlight'][field] = value
+
+    # If there is a hit in a TeX field, we prefer highlighting on that
+    # field, since other tokenizers will clobber the TeX.
+    for field in ['abstract', 'title']:
+        if f'{field}.tex' in result['highlight']:
+            result['highlight'][field] = \
+                result['highlight'].pop(f'{field}.tex')
+
+    for field in ['abstract.tex', 'abstract.english', 'abstract']:
+        if field in result['highlight']:
+            value = result['highlight'][field]
+            abstract_snippet = preview(value)
+            result['preview']['abstract'] = abstract_snippet
+            result['highlight']['abstract'] = value
+            break
+    for field in ['title.english', 'title']:
+        if field in result['highlight']:
+            result['highlight']['title'] = result['highlight'][field]
+            break
+    return result
+
+
+def _strip_highlight_and_enclose(match: Any) -> str:
+    # typing: ignore
+    value: str = match.group(0)
+    new_value = bleach.clean(value, strip=True, tags=[])
+
+    # If HTML was removed, we will assume that it was highlighting HTML.
+    if len(new_value) < len(value):
+        return f'{HIGHLIGHT_TAG_OPEN}{new_value}{HIGHLIGHT_TAG_CLOSE}'
+    return value
+
+
+def _highlight_whole_texism(value: str) -> str:
```

Comment:
This made me think of search terms that are texisms
Our highlighting highlights the numeric portions but not brackets, but that is probably fine.
https://beta.arxiv.org/search/advanced?advanced=1&terms-0-operator=AND&terms-0-term=_%7B23%7D&terms-0-field=abstract&classification-physics=y&classification-physics_archives=all&date-year=2017&date-filter_by=date_range&date-from_date=2002&date-to_date=2003&size=50&order=
Underscores are fine
https://beta.arxiv.org/search/advanced?advanced=1&terms-0-operator=AND&terms-0-term=H_20&terms-0-field=all&classification-physics_archives=all&date-filter_by=all_dates&date-year=2017&date-from_date=2002&date-to_date=2003&size=50&order=